app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,228 @@
+          end
+
+          # Creating a folder
+          dir = "#{absolute_path}/#{params[:name]}"
+          FileUtils.mkdir_p(dir)
+
+        else
+          # Uploading a file
+          input_file = params[:file]
+          return unless input_file
+          if all_filenames.include?(input_file.original_filename)
+            raise "File with name #{input_file.original_filename} already exists."
+          elsif input_file.size >= 1.gigabyte
+            raise "File size is too large. Upload a file that is smaller than 1 GB."
+          else
+            File.open(Rails.root.join(absolute_path, input_file.original_filename), 'wb') do |file|
+              file.write(input_file.read)
+            end
+          end
+        end
+      end
+    end
+  end
+
+  def my_escape(string)
+    string.gsub(/([^ a-zA-Z0-9_.-]+)/) do
+      "%#{$1.unpack('H2' * $1.bytesize).join('%').upcase}"
+    end
+  end
+
+  def populate_directory(current_directory, current_url)
+    directory = Dir.entries(current_directory)
+    new_url = current_url == '/' ? '' : current_url
+    @directory = directory.map do |file|
+      abs_path_str = "#{current_directory}/#{file}"
+      stat = File.stat(abs_path_str)
+      is_file = stat.file?
+      if %w[. ..].include?(file)
+        inst = true
+        if current_directory == BASE_DIRECTORY
+          inst = false
+        end
+      else
+        abs_path = Pathname.new(abs_path_str)
+        inst = check_instructor(abs_path)
+      end
+      {
+        size: (if is_file
+                 begin
+                   stat.size
+                 rescue StandardError
+                   '-'
+                 end
+               else
+                 '-'
+               end),
+        type: (is_file ? :file : :directory),
+        date: begin
+          stat.mtime.strftime('%d %b %Y %H:%M')
+        rescue StandardError
+          '-'
+        end,
+        relative: CGI.unescape(my_escape("/file_manager/#{new_url}#{file}")),
+        entry: "#{file}#{is_file ? '' : '/'}",
+        absolute: abs_path_str,
+        instructor: inst,
+      }
+    end.sort_by { |entry| "#{entry[:type]}#{entry[:relative]}" }
+  end
+
+  def safe_expand_path(path)
+    current_directory = Pathname.new(BASE_DIRECTORY)
+    tested_path = Pathname.new(File.join(BASE_DIRECTORY, path))
+    unless Archive.in_dir?(tested_path, current_directory, strict: false)
+      raise ArgumentError, 'Should not be parent of root'
+    end
+
+    tested_path
+  end
+
+  def check_path_exist(path)
+    @absolute_path = safe_expand_path(path)
+    @relative_path = path
+    raise ActionController::RoutingError, 'Not Found' unless File.exist?(@absolute_path)
+
+    @absolute_path
+  end
+
+  def check_instructor(path)
+    current_user_id = current_user.id
+    path = Pathname.new(path)
+    cuds = CourseUserDatum.where(user_id: current_user_id, instructor: true)
+    courses = cuds.map do |cud|
+      Course.find_by(id: cud.course_id)
```

Comment:
Nit: Instead of performing one sql query per cud, why not `courses = Course.where(id: cuds.map(&:course_id))`

(Also, the current semantics is a bit dubious because `find_by` returns `nil` if an id doesn't exist, so `find` might have been a better choice)

Reply:
Addressed.